cloudflare/data_source_zones_test.go

```diff
@@ -0,0 +1,160 @@
+package cloudflare
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/terraform/terraform"
+
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func TestAccCloudflareZonesMatchName(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCloudflareZonesConfigMatchName(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudflareZonesDataSourceID("data.cloudflare_zones.examples_domains"),
+					resource.TestCheckResourceAttr("data.cloudflare_zones.examples_domains", "zones.#", "2"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccCloudflareZonesMatchPaused(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCloudflareZonesConfigMatchPaused(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudflareZonesDataSourceID("data.cloudflare_zones.examples_domains"),
+					resource.TestCheckResourceAttr("data.cloudflare_zones.examples_domains", "zones.#", "1"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccCloudflareZonesMatchStatus(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCloudflareZonesConfigMatchStatus(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudflareZonesDataSourceID("data.cloudflare_zones.examples_domains"),
+					testAccCheckCloudflareZonesReturned("data.cloudflare_zones.examples_domains", "zones.#", func(i int) bool {
+						return i >= 1 && i <= 2
+					}),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckCloudflareZonesDataSourceID(n string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		all := s.RootModule().Resources
+		rs, ok := all[n]
+		if !ok {
+			return fmt.Errorf("Can't find zones data source: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("Snapshot zones source ID not set")
+		}
+		return nil
+	}
+}
+
+func testAccCheckCloudflareZonesReturned(n string, a string, check func(int) bool) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		all := s.RootModule().Resources
+		rs, ok := all[n]
+		if !ok {
+			return fmt.Errorf("Can't find zones data source: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("Snapshot zones source ID not set")
+		}
+
+		count, _ := strconv.Atoi(rs.Primary.Attributes[a])
+		if !check(count) {
+			return fmt.Errorf("Error evaluating %s.%s actual count: %d", n, a, count)
+		}
+		return nil
+	}
+}
+
+func testAccCloudflareZonesConfigMatchName() string {
+	return fmt.Sprintf(`
+data "cloudflare_zones" "examples_domains" {
+  filter {
+    zone   = "baa.*"
+    paused = "${cloudflare_zone.foo_net.paused}" // true
```

Comment:
**Hack** This might look a bit weird, but it's necessary to reference a resource here so it gets created before the data source look up executes.

I did try referencing the test zones from the data source using `depends_on` attribute, but this caused a test failure during the check plan diff is empty phase.